src/main/java/org/osc/manager/ism/api/IsmDeviceApi.java

```diff
@@ -62,74 +60,119 @@ public boolean isDeviceGroupSupported() {
     }
 
     @Override
     public String findDeviceByName(final String name) throws Exception {
-        
-        return txControl.supports(new Callable<String>() {
-            
+
+        return this.txControl.supports(new Callable<String>() {
+
             @Override
             public String call() throws Exception {
-                CriteriaBuilder cb = em.getCriteriaBuilder();
-                
-                CriteriaQuery<Long> q = cb.createQuery(Long.class);
-                Root<VSSDevice> r = q.from(VSSDevice.class);
-                q.select(r.get("id").as(Long.class)).where(cb.equal(r.get("name"), name));
-                
-                Long result = em.createQuery(q).getSingleResult();
-                return result == null ? null : result.toString();
+                CriteriaBuilder criteriaBuilder = IsmDeviceApi.this.em.getCriteriaBuilder();
+
+                CriteriaQuery<Long> query = criteriaBuilder.createQuery(Long.class);
+                Root<DeviceEntity> r = query.from(DeviceEntity.class);
+                query.select(r.get("id").as(Long.class)).where(criteriaBuilder.equal(r.get("name"), name));
+                List<Long> result = IsmDeviceApi.this.em.createQuery(query).getResultList();
+                return result.isEmpty() == true ? null : result.get(0).toString();
             }
         });
     }
 
     @Override
     public List<? extends ManagerDeviceElement> listDevices() throws Exception {
-        
-        return txControl.supports(new Callable<List<? extends ManagerDeviceElement>>() {
-            
+
+        return this.txControl.supports(new Callable<List<DeviceEntity>>() {
+
             @Override
-            public List<? extends ManagerDeviceElement> call() throws Exception {
-                CriteriaBuilder cb = em.getCriteriaBuilder();
-                
-                CriteriaQuery<VSSDevice> q = cb.createQuery(VSSDevice.class);
-                Root<VSSDevice> r = q.from(VSSDevice.class);
-                q.select(r);
-                
-                List<Device> devices = new ArrayList<Device>();
-                for (VSSDevice dev : em.createQuery(q).getResultList()) {
-                    devices.add(new Device(dev.getId(), dev.getName()));
-                }
-                return devices;
+            public List<DeviceEntity> call() throws Exception {
+
+                CriteriaBuilder criteriaBuilder = IsmDeviceApi.this.em.getCriteriaBuilder();
+                CriteriaQuery<DeviceEntity> query = criteriaBuilder.createQuery(DeviceEntity.class);
+                Root<DeviceEntity> r = query.from(DeviceEntity.class);
+                query.select(r);
+                return IsmDeviceApi.this.em.createQuery(query).getResultList();
             }
         });
     }
 
     @Override
     public String createVSSDevice() throws Exception {
-        return txControl.required(new Callable<VSSDevice>() {
+
+        return this.txControl.required(new Callable<DeviceEntity>() {
```

Comment:
I see you are calling this method from the rest api and because of that you need to pass a VSElement. This is a bit odd. Instead create a new method called createDevice(DeviceEntity) in this class. This method will serve as the common code between this one and the rest one. This one will create a DeviceEntity object using the vs information like in ln 127 and invoke this new common method. The rest layer already takes an entity and it will just call this common method. You can follow the same pattern across other APIs as well and this way you dont need that inner vssElement class and you still can reuse the code.

Reply:
OK